kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal;
+
+import io.delta.kernel.ScanBuilder;
+import io.delta.kernel.Snapshot;
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.Tuple2;
+
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.lang.Lazy;
+import io.delta.kernel.internal.replay.LogReplay;
+import io.delta.kernel.internal.snapshot.LogSegment;
+
+/**
+ * Implementation of {@link Snapshot}.
+ */
+public class SnapshotImpl implements Snapshot
+{
+    private final Path logPath;
+    private final Path dataPath;
+    private final long version;
+    private final LogSegment logSegment;
+    private final long timestamp;
+
+    private final LogReplay logReplay;
+    private final Lazy<Tuple2<Protocol, Metadata>> protocolAndMetadata;
+
+    public SnapshotImpl(
+        Path logPath,
+        Path dataPath,
+        long version,
+        LogSegment logSegment,
+        TableClient tableClient,
+        long timestamp)
+    {
+        this.logPath = logPath;
+        this.dataPath = dataPath;
+        this.version = version;
+        this.logSegment = logSegment;
+        this.timestamp = timestamp;
+
+        this.logReplay = new LogReplay(
+            logPath,
+            dataPath,
+            tableClient,
+            logSegment);
+        this.protocolAndMetadata = logReplay.lazyLoadProtocolAndMetadata();
+    }
+
+    @Override
+    public long getVersion(TableClient tableClient)
+    {
+        return version;
+    }
+
+    @Override
+    public StructType getSchema(TableClient tableClient)
```

Comment:
seems like `tableClient` is unused here? come back to this?

Reply:
It is not now, but in future when we make the lazy loading, this will be needed.